posts/Gunnar/2018-03-19-spec-api-modularity-patterns.adoc

```diff
@@ -0,0 +1,426 @@
+Similarly, Bean Validation providers must access field values for validating field-level constraints.
+As we've seen, the user module must open up the package with the affected types for this to work.
+For the sake of portability, the packages should not be opened up to specific implementations,
+though, but instead to the API module.
+The API module can then open packages from the user's module also to spec implementation modules.
+This approach achieves the goal of portability between specs,
+while enabling spec implementations to perform the required private access operations.
+
+Note that this requires Java specifications to define a module name to be used by all API modules created for this specification (_fieldreader.spec_ in this example).
+Or, better yet, just provide a single API module which is shared by all the implementations.
+As an example, that already is the case for JPA as of version 2.2,
+where Hibernate ORM also depends on the _javax.persistence:javax.persistence-api:2.2_ artifact instead of providing its own artifact with the spec packages.
+
+An alternative approach could be to have the user code itself bootstrap a lookup with private access and pass this into the API during bootstrap.
+For several reason's that's not ideal, though: the `Lookup` API is rather low-level and user code might fail to correctly implement retrieval of a lookup with private access.
+Also, in container environments, the user code isn't in control of bootstrapping APIs such as JPA or Bean Validation and thus cannot pass a lookup object.
+Finally, if the user classes are spread out across multiple modules, it'd be a challenging task to collect the required lookups from all these user modules.
+
+In comparison, the solution of opening up user packages to spec API modules
+and let those further open the user packages to implementation modules is much easier to work with.
+
+Some adjustments to spec API modules will be needed for this.
+It will be interesting to see when the affected specifications will begin to explore the required changes and release updated revisions of their APIs as JPMS-aware modules.
+With the recent move of Java EE to the Eclipse Foundation and its https://mmilinkov.wordpress.com/2018/02/26/and-the-name-is/[rebranding as Jakarta EE],
+there's a great opportunity for the community - i.e. you - to help with that!
+
+You can find the complete source code for this blog post including all the three modules we've discussed in our https://github.com/hibernate/hibernate-demos/tree/master/java9/spec-modularity-patterns[examples repository].
+
+In a follow-up post we'll explore how spec implementations can deal with resources provided in application modules,
+for instance XML descriptors such as JPA's _persistence.xml_ file.
+If you got any feedback or suggestions for other related matters to discuss, let us know in the comments below.
```

Comment:
I don't know about you, but I don't get any notification when someone adds a comment to my posts. You might want to provide another way to reach out to you.

Reply:
Yes, it's a pain. I keep an eye on posts for a while after I published them, but it's not satisfying.